src/sparseml/pytorch/sparsification/pruning/modifier_pruning_obs.py

```diff
@@ -0,0 +1,580 @@
+                            * (mask_diffs[i] == -1).view(-1).to(self._devices[i])
+                        )
+                        .view(param.data.shape)
+                    )
+
+        self._broadcast_list_from_main(obs_updates)
+        # apply OBS update and manually zero-out pruned weights
+        for i, param in enumerate(self._params):
+            param.data -= obs_updates[i].to(param.data.device)
+            param.data[mask_diffs[i] == -1] = 0.0
+
+        self._finvs = None
+
+    def _validate(self):
+        if self._mask_type == "block4":
+            for param in self._params:
+                if param.numel() % self._fisher_block_size != 0:
+                    raise ValueError(
+                        "number of elements in each param must be divisible \
+                        by fisher_block_size"
+                    )
+
+    def _setup_fisher_inverse(self, masks: List[Tensor]):
+        self._masks = masks  # to be used by score_parameters
+        self._finvs = []
+        for i, param in enumerate(self._params):
+            self._finvs.append(
+                EmpiricalBlockFisherInverse(
+                    self._num_grads,
+                    self._fisher_block_size,
+                    param.numel(),
+                    self._damp,
+                    self._devices[i],
+                )
+            )
+
+
+class EmpiricalBlockFisherInverse:
+    def __init__(
+        self,
+        num_grads: int,
+        fisher_block_size: int,
+        num_weights: int,
+        damp: float,
+        device: torch.device,
+    ):
+        self.m = num_grads
+        self.B = fisher_block_size
+        self.d = num_weights
+        self.damp = damp
+        self.dev = device
+
+        self.num_blocks = math.ceil(self.d / self.B)
+        self.f_inv = (
+            (1.0 / self.damp * torch.eye(n=self.B, device=self.dev))
+            .unsqueeze(0)
+            .repeat(self.num_blocks, 1, 1)
+        )  # O(d x B) memory
+
+    def add_grad(self, g: Tensor):
+        """
+        Updates empirical Fisher inverse with a new gradient
+        :param g: a collected gradient
+        """
+        # if 'd / B' is not integer, pad with zeros for batch calculations
+        if g.numel() < self.num_blocks * self.B:
+            g = torch.cat(
+                [g, torch.zeros(self.num_blocks * self.B - g.numel(), device=g.device)]
+            )
+
+        # prepare grad for batch calculations
+        g = g.view(self.num_blocks, self.B)
+
+        # batched f_inv x g: (batch, B, B) x (batch, B) -> (batch, B)
+        finv_g = torch.einsum("bij,bj->bi", self.f_inv, g)
+
+        # scalar denominator for each batch: (batch)
+        alpha = (self.m + torch.einsum("bi,bi->b", g, finv_g)).sqrt().unsqueeze(1)
+        finv_g /= alpha
+
+        # update f_inv with new outer product: (batch, B) x (batch, B) -> (batch, B, B)
+        self.f_inv.baddbmm_(finv_g.unsqueeze(2), finv_g.unsqueeze(1), alpha=-1)
```

Comment:
Can we change the method name `baddbmm` to something more verbose?

Reply:
`baddbmm` is a method from PyTorch https://pytorch.org/docs/stable/generated/torch.baddbmm.html